### DISCUSSION TITLE
import.meta.env and process.env not working in App.tsx

### DISCUSSION THREAD

**Discussion by basit-qayoom** [web-infra-dev/rsbuild#3290] [Q&A]
# `import.meta.env` and `process.env` not working in App.tsx

## Description
I'm encountering issues while trying to access environment variables in my `App.tsx` file.

1. When I try to console `import.meta.env`, I receive the following TypeScript error:
Property ‘env’ does not exist on type ‘ImportMeta’.ts(2339)

2. When attempting to console `process.env`, the browser shows:
process is not defined

## Steps to Reproduce
1. Create an `.env` file in the root directory.
2. In `App.tsx`, add the following code:
```typescript
console.log(import.meta.env);
console.log(process.env);

3. Observe the TypeScript error and the runtime error in the browser.

```{
  "react": "^18.3.1",
  "react-dom": "^18.3.1",
  "@rsbuild/core": "1.0.1-beta.15",
  "@rsbuild/plugin-react": "1.0.1-beta.15"
} ```

Expected Behavior

I expect import.meta.env to log the environment variables defined in the .env file without any TypeScript errors. Similarly, process.env should log the environment variables without causing a runtime error.

Actual Behavior

	•	TypeScript throws an error when accessing import.meta.env.
	•	The browser throws an error when accessing process.env.

Additional Context
The .env file is placed in the root directory, and I have already ensured it is correctly formatted.

**Comment by chenjiahan** [comment]
You should always use `import.meta.env.SOME_NAME` or `process.env.SOME_NAME`.

see https://rsbuild.dev/guide/advanced/env-vars#identifiers-matching

**Comment by abenhamdine** [comment]
also see https://github.com/web-infra-dev/rsbuild/issues/3029 for more background and history

**Comment by RoyRao2333** [comment]
@chenjiahan 

It's not working with the full name on my side:

```
// .env.local

ABC=1
```

```
// App.tsx

console.log(import.meta.env.ABC)
```

Then I got error:

<img width="450" alt="image" src="https://github.com/user-attachments/assets/30df9f89-ae58-440b-a8a1-ebc099c127f9">

But I can successfully do `console.log(import.meta.env.MODE);`:

<img width="336" alt="image" src="https://github.com/user-attachments/assets/b0293bfe-43f5-43c0-9767-2be41b8a7c96">

**Reply by chenjiahan** [reply]
See "Public Variables": https://rsbuild.dev/guide/advanced/env-vars#public-variables

**Reply by RoyRao2333** [reply]
@chenjiahan 

My info may be sensitive, so it may not be a PUBLIC variable.

<img width="930" alt="image" src="https://github.com/user-attachments/assets/5374d89a-9813-4857-ab90-5c642fba66f3">

So in this case, how can I access PASSWORD in my code? Thanks!

**Reply by chenjiahan** [reply]
You should not access private variable in the client code

**Reply by RoyRao2333** [reply]
@chenjiahan 

I'm deploying my site on Vercel, can I access private variables in Serverless Functions in `api/**`?

**Reply by chenjiahan** [reply]
I am not familiar with Vercel, maybe you need to check Vercel's documentation

**Reply by RoyRao2333** [reply]
@chenjiahan 

感谢大佬的解释，懒得打英文了，我换一个描述形式吧：

本人学艺不精，这是我的理解：RsBuild作为前端项目的构建工具，那么理论上构建的代码都算client代码，那么举个例子，如何在React项目中访问私有环境变量呢？如果不能访问的话，那么私有环境变量存在的意义是什么呢？感谢！

**Reply by chenjiahan** [reply]
Are you using Rsbuild to build the `api/**` modules?

**Reply by RoyRao2333** [reply]
@chenjiahan 

`src/api/**` is under src, so it might be built with RsBuild?

**Reply by chenjiahan** [reply]
If you are sure that the module is built by Rsbuild, you can use `source.define` to replace the environment variables in it
